luaui/Widgets/gui_top_bar.lua

```diff
@@ -2054,22 +3224,125 @@ function shutdown()
+					active = true
+					if commands[i].id < 0 and not builtUnitDefID then
+						-- We're building something. A negative ID is the unitdef ID of what's being built.
+						builtUnitDefID = -commands[i].id
+						--Spring.Echo("unit " .. unitID .. " is building unitdef " .. builtUnitDefID)
+					elseif commands[i].id == CMD.GUARD then
+						mayBeBuilding = true -- building can be caused by a guard command
+						guardedUnitID = commands[i].params[1]
+					elseif commands[i].id == CMD.REPAIR then
+						mayBeBuilding = true -- building can be caused by a repair command
+					end
+				end
+			end
+		end
+	end
+
+	return unitExists, active, builtUnitDefID, mayBeBuilding, guardedUnitID
 end
 
-function widget:SetConfigData(data)
-	if data.autoHideButtons ~= nil then
-		autoHideButtons = data.autoHideButtons
+function pid(Kp, Ki, Kd, dt, prevIntegral, prevError, setPoint, measuredValue)
+	if dt <= 0 then
+		return measuredValue, prevIntegral, prevError
+	end
+	-- setpoint: instantaneous observed value
+	-- measuredValue: position of displayed indicator
+	local error = setPoint - measuredValue
+	local proportional = error;
+	local integral = prevIntegral + error * dt
+	local derivative = (error - prevError) / dt
+	local output = Kp * proportional + Ki * integral + Kd * derivative
+	return output, integral, error
+end
+
+function TrackUnit(unitID, unitDefID, unitTeam) --needed for exact calculations
+	if (myTeamID == unitTeam) then
+		local unitDef = UnitDefs[unitDefID]
+		local _, _, _, _, build = Spring.GetUnitHealth(unitID) --is it finished?
+		local isBuilt = build >= 1
+		if isBuilt and unitDef then
+			if unitDef.buildSpeed and unitDef.buildSpeed > 0 and unitDef.canAssist or UnitDefs[unitDefID].isFactory then
+				trackedNum = trackedNum + 1
+				if not trackedBuilders[unitID] then				-- We may have already tracked this unit when it started being built. No need to add its BP again.
+					BP[4] = BP[4] + unitDef.buildSpeed -- BP[4] ^= totalAvailableBP
+				end
+				trackedBuilders[unitID] = { unitDef.buildSpeed, isBuilt, unitDefID, unitTeam }
+			elseif unitDef.name == "armwin" or unitDef.name == "corwin" then -- wind generator
+				trackedWinds[unitID] = 1
+				numWindGenerators = numWindGenerators + 1
+			end
+		end
+	end
+end
+
+function UntrackUnit(unitID, unitDefID, unitTeam) -- needed for exact calculations
+	local unitDef = UnitDefs[unitDefID]
+	if (myTeamID == unitTeam) and unitDef then
+		if trackedBuilders[unitID] then
+			if unitDef.buildSpeed and unitDef.buildSpeed > 0 and unitDef.canAssist or UnitDefs[unitDefID].isFactory then
+			trackedNum = trackedNum - 1
+				BP[4] = BP[4] - trackedBuilders[unitID][1] -- BP[4] ^= totalAvailableBP
+			end
+		end
+		if trackedBuilders[unitID] then
+			 trackedBuilders[unitID] = nil
+		end
+		if trackedWinds[unitID] then
+			trackedWinds[unitID] = nil
+			numWindGenerators = numWindGenerators - 1
+		end
+	end
+end
+
+function InitAditionalValues()
+	BP[4] = 0
+	trackedNum = 0
```

Comment:
`InitAditionalValues` is only called in `Initialize`, so whats the benefit of setting `trackedNum` here rather than #L283?

Reply:
The widget needs to InitAditionalValues when players change, too.


function widget:PlayerChanged()
	local prevSpec = spec
	spec = spGetSpectatingState()
	checkSelfStatus()
	numTeamsInAllyTeam = #Spring.GetTeamList(myAllyTeamID)
	InitAllUnits()
	
InitAllUnits wasn't renamed.

Done. Will upload, soon.

Reply:
fixed version is online